dicomparser/parsers.bal

```diff
@@ -0,0 +1,368 @@
+    } on fail error e {
+        return error dicom:ParsingError(string `VR parsing failed: ${vrBytes.toString()}`, e);
+    }
+}
+
+# Parses a DICOM Value Length (VL).
+#
+# + vlBytes - The byte array containing the encoded VL
+# + byteOrder - The byte order used in the byte array
+# + return - The `int` representation of the VL, or a `dicom:ParsingError` if the parsing fails.
+public isolated function parseVl(byte[] vlBytes, dicom:ByteOrder byteOrder) returns int|dicom:ParsingError {
+    do {
+        // VL must be either 2 or 4 bytes in length
+        if vlBytes.length() != 2 && vlBytes.length() != 4 {
+            fail error("VL bytes must be of length 2 or 4");
+        }
+        return check dicom:bytesToInt(vlBytes, byteOrder);
+    } on fail error e {
+        return error dicom:ParsingError(string `VL parsing failed: ${vlBytes.toString()}`, e);
+    }
+}
+
+# Parses a DICOM data element value.
+#
+# + vr - The Value Representation (VR) of the data element
+# + valueBytes - The byte array containing the encoded data element value
+# + byteOrder - The byte order used in the byte array
+# + return - A `dicom:DataElementValue` if the parsing is successful, or a `dicom:ParsingError` if the parsing fails
+public isolated function parseValue(dicom:Vr vr, byte[] valueBytes,
+        dicom:ByteOrder byteOrder) returns dicom:DataElementValue|dicom:ParsingError {
+    do {
+        // Value parsing logic is based off of Table 6.2-1 in Part 5
+        dicom:DataElementValue value = ();
+
+        // If VR has a fixed value length bytes, and the valueBytes[] length does not match the length 
+        // in the spec, don't try to parse, use valueBytes[] as the value
+        if dicom:FIXED_LENGTH_VALUE_BYTES.hasKey(vr) {
+            if dicom:FIXED_LENGTH_VALUE_BYTES.get(vr) != valueBytes.length() {
+                value = valueBytes;
+                return value;
+            }
+        }
+
+        if vr == dicom:AT {
+            // AT is an ordered pair of 16-bit unsigned integers that is the value of a data element tag
+            value = check parseTag(valueBytes, byteOrder);
+        }
+
+        if dicom:isBytesVr(vr) {
+            value = valueBytes;
+        }
+
+        if dicom:isFloatVr(vr) {
+            value = check dicom:bytesToFloat(valueBytes, byteOrder);
+        }
+
+        if dicom:isIntVr(vr) {
+            value = check dicom:bytesToInt(valueBytes, byteOrder);
+        }
+
+        if dicom:isStringVr(vr) {
+            value = check string:fromBytes(valueBytes);
+        }
+
+        // Remove leading and trailing spaces if insignificant
+        if dicom:SPACE_INSIGNIFICANT_VRs.indexOf(vr) != () {
+            if value is string {
+                value = value.trim();
+            }
+        }
+
+        // If the value is of string type
+        // Remove leading and trailing spaces
+        if value is string {
+            value = value.trim();
+        }
```

Comment:
These subsequent checks as a chain with else-if ? Seems like unnecessary checking for all the cases if one of them satisfied on evaluation

Reply:
This was due to an issue with some VRs belonging to two data types. Therefore, had to be processed like this (more than once). This issue was fixed in  https://github.com/ballerina-platform/module-ballerinax-health.dicom/commit/3807383514d097fe372c936aa46ef84a627a58d3 and these checks now use a chain of if-else-if statements.